### PR TITLE
Make ObjectMapper of CustomResourceCache configurable

### PR DESCRIPTION
**Pull request by i8r**
* Optionally provide the ObjectMapper in the constructor of the Operator
* Optionally provide the ObjectMapper in constructor of the CustomResourceCache
* Optionally autowire a ObjectMapper in spring boot

### CODE REVIEW

**Comment by i8r** [comment]
Addressing #385

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java [hunk 1]
```diff
@@ -3,42 +3,38 @@
...
 
   public void cacheResource(CustomResource resource, Predicate<CustomResource> predicate) {
-    try {
-      lock.lock();
-      if (predicate.test(resources.get(KubernetesResourceUtils.getUID(resource)))) {
-        log.trace("Update cache after condition is true: {}", resource);
-        resources.put(resource.getMetadata().getUid(), resource);
+    resources.compute(KubernetesResourceUtils.getUID(resource), (id, prev) -> {
+      if (predicate.test(prev)) {
+        log.trace("Attempting to update cache after condition is true: {}", resource);
```
Comment (i8r) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java:33]: This log became a little less precise but is it actually needed?

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java [hunk 2]
```diff
@@ -3,42 +3,39 @@
...
 
   private static final Logger log = LoggerFactory.getLogger(CustomResourceCache.class);
 
-  private ObjectMapper objectMapper = new ObjectMapper();
-  private final Map<String, CustomResource> resources = new ConcurrentHashMap<>();
-  private final Lock lock = new ReentrantLock();
+  private final ObjectMapper objectMapper;
+  private final ConcurrentMap<String, CustomResource> resources = new ConcurrentHashMap<>();
+
+  public CustomResourceCache(ObjectMapper objectMapper) {
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java:21]: I would add a default no-arg constructor that would just use the default ObjectMapper. This way, existing code wouldn't have to be changed.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java [hunk 3]
```diff
@@ -3,42 +3,39 @@
...
-    try {
-      lock.lock();
-      if (predicate.test(resources.get(KubernetesResourceUtils.getUID(resource)))) {
-        log.trace("Update cache after condition is true: {}", resource);
-        resources.put(resource.getMetadata().getUid(), resource);
-      }
-    } finally {
-      lock.unlock();
-    }
+    resources.compute(
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java:30]: This change seems unrelated to the PR's focus, especially considering they change the access pattern to the underlying structure so I'd rather not see these changes in this PR. Note that there is already an open issue to replace this cache implementation by a more appropriate one (#356).

**Review by metacosm — CHANGES_REQUESTED** [review]
Please use conventional commits (https://www.conventionalcommits.org/en/v1.0.0/) for commit messages.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java (hunk 2 above)
Comment (i8r) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java:28]: Sure, admittedly it is a public class though I don't see why you would use it outside of the library itself.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java (hunk 3 above)
Comment (i8r) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java]: I've undone the changes. They are definitely unrelated. I just tried to apply the boy scout rule.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java [hunk 4]
```diff
@@ -49,18 +58,13 @@ public void cacheResource(CustomResource resource, Predicate<CustomResource> pre
    * @return
    */
   public Optional<CustomResource> getLatestResource(String uuid) {
-    return Optional.ofNullable(clone(resources.get(uuid)));
+    return Optional.ofNullable(resources.get(uuid)).map(this::clone);
```
Comment (i8r) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java:61]: Strictly speaking this is still unrelated. I'll undo it as well if you prefer.

**Comment by i8r** [comment]
Comments address and commit history cleaned up

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java (hunk 2 above)
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java:28]: This would allow to leave the test classes untouched, for example… 😉

File: operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/DefaultEventHandlerTest.java [hunk 5]
```diff
@@ -36,7 +37,7 @@
   public static final int FAKE_CONTROLLER_EXECUTION_DURATION = 250;
   public static final int SEPARATE_EXECUTION_TIMEOUT = 450;
   private EventDispatcher eventDispatcherMock = mock(EventDispatcher.class);
-  private CustomResourceCache customResourceCache = new CustomResourceCache();
+  private CustomResourceCache customResourceCache = new CustomResourceCache(new ObjectMapper());
```
Comment (metacosm) [inline operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/DefaultEventHandlerTest.java:40]:
```suggestion
  private CustomResourceCache customResourceCache = new CustomResourceCache();
```

File: operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/event/internal/CustomResourceEventSourceTest.java [hunk 6]
```diff
@@ -19,7 +20,7 @@
 class CustomResourceEventSourceTest {
 
   public static final String FINALIZER = "finalizer";
-  CustomResourceCache customResourceCache = new CustomResourceCache();
+  CustomResourceCache customResourceCache = new CustomResourceCache(new ObjectMapper());
```
Comment (metacosm) [inline operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/event/internal/CustomResourceEventSourceTest.java:23]:
```suggestion
  CustomResourceCache customResourceCache = new CustomResourceCache();
```

File: operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/event/internal/CustomResourceEventSourceTest.java [hunk 7]
```diff
@@ -5,6 +5,7 @@
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Comment (metacosm) [inline operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/event/internal/CustomResourceEventSourceTest.java:8]:
```suggestion
```

File: operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/DefaultEventHandlerTest.java [hunk 8]
```diff
@@ -11,6 +11,7 @@
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Comment (metacosm) [inline operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/DefaultEventHandlerTest.java:14]:
```suggestion
```

**Comment by metacosm** [comment]
Can you please fix the typo in the commit message? Otherwise, this looks good to me.

**Comment by i8r** [comment]
> Can you please fix the typo in the commit message? Otherwise, this looks good to me.

Of course. 

I also addressed the other comments.

**Comment by i8r** [comment]
> Sorry to be a bother but now the commit message doesn't follow the conventional commit format anymore… 😢

Oh no, you are right. Sorry for messing up the commit message multiple times 😄

**Review by metacosm — APPROVED** [review]

**Comment by metacosm** [comment]
Thank you!
